jaeger-core/src/test/java/io/jaegertracing/internal/JaegerSubclassTest.java

```diff
@@ -0,0 +1,179 @@
+  }
+
+  private static class Tracer extends JaegerTracer {
+    public class SpanBuilder extends JaegerTracer.SpanBuilder {
+      protected SpanBuilder(String operationName) {
+        super(operationName);
+      }
+
+      @Override
+      public Span start() {
+        return new Span(super.start());
+      }
+
+      @Override
+      public Scope startActive(boolean finishSpanOnClose) {
+        return scopeManager().activate(start(), finishSpanOnClose);
+      }
+    }
+
+    Tracer(io.jaegertracing.internal.JaegerTracer tracer) {
+      this(
+          tracer.getServiceName(),
+          tracer.getReporter(),
+          tracer.getSampler(),
+          tracer.getRegistry(),
+          tracer.clock(),
+          tracer.getMetrics(),
+          new HashMap<>(tracer.tags()),
+          tracer.isZipkinSharedRpcSpan(),
+          tracer.scopeManager(),
+          tracer.getBaggageRestrictionManager(),
+          tracer.isExpandExceptionLogs());
+    }
+
+    public Tracer(
+        String serviceName,
+        Reporter reporter,
+        Sampler sampler,
+        PropagationRegistry registry,
+        Clock clock,
+        Metrics metrics,
+        Map<String, Object> tags,
+        boolean zipkinSharedRpcSpan,
+        ScopeManager scopeManager,
+        BaggageRestrictionManager baggageRestrictionManager,
+        boolean expandExceptionLogs) {
+      super(
+          serviceName,
+          reporter,
+          sampler,
+          registry,
+          clock,
+          metrics,
+          tags,
+          zipkinSharedRpcSpan,
+          scopeManager,
+          baggageRestrictionManager,
+          expandExceptionLogs);
+    }
+
+    @Override
+    public SpanBuilder buildSpan(String operationName) {
+      return new SpanBuilder(operationName);
+    }
+  }
+
+  private static class Span extends JaegerSpan {
+    private final SpanContext ctx;
+
+    Span(JaegerSpan span) {
+      this(
+          span.getTracer(),
+          span.getOperationName(),
+          new SpanContext(span.context()),
+          span.getStart(),
+          span.getStartTimeNanoTicks(),
+          span.getComputeDurationViaNanoTicks(),
+          span.getTags(),
+          span.getReferences());
+    }
+
+    public Span(
+        JaegerTracer tracer,
+        String operationName,
+        JaegerSpanContext context,
+        long startTimeMicroseconds,
+        long startTimeNanoTicks,
+        boolean computeDurationViaNanoTicks,
+        Map<String, Object> tags,
+        List<Reference> references) {
+      super(
+          tracer,
+          operationName,
+          context,
+          startTimeMicroseconds,
+          startTimeNanoTicks,
+          computeDurationViaNanoTicks,
+          tags,
+          references);
+      ctx = new SpanContext(context);
+    }
+
+    @Override
+    public SpanContext context() {
+      return ctx;
+    }
+  }
+
+  private static class SpanContext extends JaegerSpanContext {
+    SpanContext(JaegerSpanContext ctx) {
```

Comment:
That's not what I thought you were going after. If you extend the context, why do you need to accept it as agrument? The oss tracer currently does "new SpanContext". You can move that call into a function that you would override in the subclassed tracer and return your own subclass of SpanContext.

Reply:
I could redo a lot of code. The idea is to avoid code duplication by allowing base classes to do 99% of the work. Imagine the case of a JaegerTracer starting a JaegerSpan. There is a lot of logic dedicated to determining the span parent, etc. Now I could override `start` because Tracer needs to return `Span` instead of `JaegerSpan`, but that's going to result in a huge amount of copy-paste. Isn't it easier to use a constructor `Span(JaegerSpan span)` (note package visibility, not public)?

Reply:
I am not suggesting redoing any code. What I am saying is that, for example, Trace has this code in `builder.createNewContext()`:
```
return new JaegerSpanContext(id, id, 0, flags);
```

`createNewContext` itself is a large method we don't want to replicate. But we can change the above line to
```
return instantiateNewSpanContext(id, id, 0, flags);
```

and add method:
```
protected JaegerSpanContext instantiateNewSpanContext(...) {
  return new JaegerSpanContext(id, id, 0, flags);
}
```

Then in your subclassed Tracer you just need to override this method
```
@Override
protected JaegerSpanContext instantiateNewSpanContext(...) {
  return new SubclassedJaegerSpanContext(id, id, 0, flags);
}
```

Your subclassed tracer will have a total of 3 methods: constructor that directly delegates to `super(....same args.....)`, and two instantiate methods, for span context and the Span that create 
subclassed instances.

Reply:
OK now I understand. I think this is called the [abstract factory pattern](https://en.wikipedia.org/wiki/Abstract_factory_pattern). That definitely makes sense. I am usually hesitant to change so much of the existing code, but I guess this is a better design.